
syntax = 'proto3';


/**
 * Collect mode
 */
enum CollectMode {
  UNKNOWN_MODE = 0;
  PACKET = 1;
  PACKET_ID = 2;
}


/**
 * metadata for Netfilter
 */
message NetFilterInfo {
  string hookpoint = 1;
}


message eBPFInfo {
  string hookpoint = 1;
}


/**
 * packet
 */
message PacketInfo {
  oneof data {
   // packet byte data
    bytes packet = 1;
    uint64 packet_id = 2;
  }
  // node name (id)
  uint32 node_id = 5;
  // packet sniff timestamp
  double timestamp = 6;

  oneof metadata {
    NetFilterInfo netfilterInfo = 10;
    eBPFInfo ebpfInfo = 11;
  }
}

/**
 * Agent setting
 */
message PollSettingRequest {
  uint32 node_id = 2;
  uint32 node_id_length = 3;
  uint32 counter_length = 4;
}

/**
 * Setting Reply
 */
message PollSettingReply {}

/**
 * Packet Request
 */
message PacketInfoRequest {}

message PacketInfoReply {
  repeated PacketInfo packet_infos = 1;
}


/**
 * NotifyPacket Request
 */
message PacketInfoStreamRequest {
  optional bool capture_all_packets = 1;
  uint32 node_id = 2;
  uint32 node_id_length = 3;
  uint32 counter_length = 4;
}


/**
 * Packet Collector
 */
service PacketCollectService {
  rpc SetPoll(PollSettingRequest) returns (PollSettingReply) {}
  rpc GetPacketInfo(PacketInfoRequest) returns (PacketInfoReply) {}
  rpc GetPacketInfoStream(PacketInfoStreamRequest) returns (stream PacketInfo) {}
}